lib/matplotlib/backends/backend_mixed.py

```diff
@@ -84,24 +85,36 @@ def start_rasterizing(self):
             r = process_figure_for_rasterizing(self.figure,
                                                self._bbox_inches_restore)
             self._bbox_inches_restore = r
-        if self._rasterizing == 0:
+        if self._rasterizing == 0 and not self._rasterizing_unflushed:
             self._raster_renderer = self._raster_renderer_class(
                 self._width*self.dpi, self._height*self.dpi, self.dpi)
             self._renderer = self._raster_renderer
         self._rasterizing += 1
+        self._rasterizing_unflushed = True
 
     def stop_rasterizing(self):
         """
-        Exit "raster" mode.  All of the drawing that was done since
-        the last `start_rasterizing` call will be copied to the
-        vector backend by calling draw_image.
+        Exit "raster" mode. The accumulated bitmap will not be copied to
+        the output until `flush_rasterizing` is called.
 
         If `start_rasterizing` has been called multiple times,
         `stop_rasterizing` must be called the same number of times before
         "raster" mode is exited.
         """
         self._rasterizing -= 1
-        if self._rasterizing == 0:
+
+        if self._bbox_inches_restore:  # when tight bbox is used
+            r = process_figure_for_rasterizing(self.figure,
+                                               self._bbox_inches_restore,
+                                               self._figdpi)
+            self._bbox_inches_restore = r
+
+    def flush_rasterizing(self):
+        """
+        All of the drawing that was done since the last `start_rasterizing`
+        call will be copied to the vector backend by calling draw_image.
+        """
+        if self._rasterizing == 0 and self._rasterizing_unflushed:
```

Comment:
Is this defensive programming or does this regularly get called "out of sequence"?

I think we should at least warn (and maybe raise?) if this is being called "out of sequence" as that feels like an internal programming error to me.

There should also be some thought about how to clean this state up if something goes wrong during the rendering.  In addition to the call in the `finally` in the wrapper, this block should probably be in a `try...finally` that resets the flushed state.

Reply:
Ah :sheep: I now see that we need to call this at the end of `Figure.draw` and it is better to call it un-conditionally.

Reply:
Its due to to the sequencing. We need to flush out the bitmap before drawing a vector part or at the end of drawing the whole figure, but only if there is actually something to flush.

Also, if we draw and artist that is does not have rasterized set, but is a child of a one that does, then we stay in raster mode. Hence the 2 tests.